### PR TITLE
fix: apply cert-manager default issuer kind/group when matching policies

### PR DESCRIPTION
**Pull request by erikgb**
This PR makes approver-policy adhere to the cert-manager controller defaults for issuer `kind`/`group`.

When referencing issuers in cert-manager `Certificate` and `CertificateRequest`, the issuerRef `kind` and `group` are optional and defaulted in the cert-manager controller (a.k.a. controller defaults). This becomes problematic in approver-policy if you want to enforce a policy addressing cert-manager issuers. Example (irrelevant details omitted):

Policy:
````yaml
apiVersion: policy.cert-manager.io/v1alpha1
kind: CertificateRequestPolicy
metadata:
  name: allow-all-cert-manager-issuers
spec:
  selector:
    issuerRef:
      group: cert-manager.io
      kind: Issuer
````

Certificate:
````yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: my-cert
spec:
  issuerRef:
    name: my-issuer
````

This kind of simple certificate is common in operators using kubebuilder/operator-sdk for development. The reason probably being to keep the scaffolding in these CLI tools as simple as possible,

Expected behavior: The policy matches certificate requests because cert-manager defaults match policy issuerRef `group` and `kind`.

Actual behavior: Policy doesn't select the requests ending up in never-approved requests. 😒 

Possible workaround: Use a mutating webhook to materialize the cert-manager controller defaults into cert-manager resources. Tutorial: https://cert-manager.io/docs/tutorials/certificate-defaults/. I am personally not a big fan of this workaround, as every webhook call increases the operational risk of outages in your cluster.

This issue was also discussed in this long thread on [Slack](https://kubernetes.slack.com/archives/C4NV3DWUC/p1723537145338019), even if it turned out to not be the root cause of the issue that started the thread.

### CODE REVIEW

File: pkg/internal/approver/manager/predicate/predicate.go
Comment (wallrj) [inline pkg/internal/approver/manager/predicate/predicate.go:78]: I was going to suggest using `scheme.Default(cr)` but that doesn't work, because cert-manager doesn't encapsulate all the resource defaults in the API...I wish it did.

```go
	s := runtime.NewScheme()
	if err := cmapi.AddToScheme(s); err != nil {
		return nil, err
	}
	s.Default(cr)
```

* https://github.com/cert-manager/cert-manager/blob/master/internal/apis/certmanager/v1/zz_generated.defaults.go

**Review by wallrj — APPROVED** [review]
Thanks @erikgb 

I ran the new unit tests and saw it fail with the previous code and pass with your updated code;


Before:
```console
$ go test ./pkg/internal/approver/manager/predicate/... -run 'Test_SelectorIssuerRef'
--- FAIL: Test_SelectorIssuerRef (0.00s)
    --- FAIL: Test_SelectorIssuerRef/if_policy_specifies_cert-manager_defaults_and_request_omits_defaults,_return_policy (0.00s)
        predicate_test.go:728: unexpected policies returned:
            exp=[]v1alpha1.CertificateRequestPolicy{v1alpha1.CertificateRequestPolicy{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1alpha1.CertificateRequestPolicySpec{Allowed:(*v1alpha1.CertificateRequestPolicyAllowed)(nil), Constraints:(*v1alpha1.CertificateRequestPolicyConstraints)(nil), Plugins:map[string]v1alpha1.CertificateRequestPolicyPluginData(nil), Selector:v1alpha1.CertificateRequestPolicySelector{IssuerRef:(*v1alpha1.CertificateRequestPolicySelectorIssuerRef)(0xc0001394a0), Namespace:(*v1alpha1.CertificateRequestPolicySelectorNamespace)(nil)}}, Status:v1alpha1.CertificateRequestPolicyStatus{Conditions:[]v1alpha1.CertificateRequestPolicyCondition(nil)}}}
            got=[]v1alpha1.CertificateRequestPolicy(nil)
FAIL
FAIL    github.com/cert-manager/approver-policy/pkg/internal/approver/manager/predicate 0.054s
FAIL
```

After:

```console
$ go test ./pkg/internal/approver/manager/predicate/... -run 'Test_SelectorIssuerRef'
ok      github.com/cert-manager/approver-policy/pkg/internal/approver/manager/predicate 0.025s
```

I haven't done any E2E testing, but I trust that you have.

I can't see any backwards compatibility issues. Anyone who might be using only an external issuer which reconciles CRs without default group and kind, might I suppose rely on this current behaviour...but can't imagine anyone doing that.

/approve
/lgtm
/hold in case you want to make the new string defaulting closure a normal function.

File: pkg/internal/approver/manager/predicate/predicate.go
Comment (erikgb) [reply pkg/internal/approver/manager/predicate/predicate.go:68]: I prefer the closure here. Naming a simple function as this is hard, ref. the closed Go issue.

**Comment by erikgb** [comment]
> I haven't done any E2E testing, but I trust that you have.

@wallrj I haven't been able to test this end-to-end. I would love to write an e2e-test for this, but I don't want to duplicate a lot of code. Could you please take a look at https://github.com/cert-manager/approver-policy/pull/525?

I am putting this PR back into draft mode while I work on adding an e2e-test.

File: pkg/internal/approver/manager/predicate/predicate.go
Comment (erikgb) [reply pkg/internal/approver/manager/predicate/predicate.go]: After thinking a bit, I agree with you. 😺 I have replaced the closure with a normal function now.

File: pkg/internal/approver/manager/predicate/predicate.go
Comment (erikgb) [reply pkg/internal/approver/manager/predicate/predicate.go:71]: Like this maybe, @wallrj? 😉 https://github.com/cert-manager/cert-manager/pull/7414/files

**Comment by erikgb** [comment]
> > I haven't done any E2E testing, but I trust that you have.
> 
> @wallrj I haven't been able to test this end-to-end. I would love to write an e2e-test for this, but I don't want to duplicate a lot of code. Could you please take a look at #525?
> 
> I am putting this PR back into draft mode while I work on adding an e2e-test.

@wallrj After our common understanding in today's stand-up, I added an integration test (where envTest is used) instead of an additional e2e-test running on kind. Should be ready for another round of review now. PTAL!

File: pkg/internal/approver/manager/predicate/predicate_test.go
Comment (wallrj) [reply pkg/internal/approver/manager/predicate/predicate_test.go:548]: There's no way to focus the tests in this repo, so I resorted to the following patch

```diff

```diff
diff --git a/make/test-unit.mk b/make/test-unit.mk
index 7180200..a8fbb2d 100644
--- a/make/test-unit.mk
+++ b/make/test-unit.mk
@@ -30,7 +30,7 @@ test-unit: | $(cert_manager_crds) $(approver_policy_crds) $(NEEDS_GO) $(NEEDS_GI
                --junit-report=junit-go-e2e.xml \
                --cover \
                --coverprofile=filtered.cov \
-               ./cmd/... ./pkg/... \
+               ./pkg/internal/approver/manager/predicate/... \
                -ldflags $(go_manager_ldflags)

        $(GO) tool cover -html=$(ARTIFACTS)/filtered.cov -o=$(ARTIFACTS)/filtered.html
```

```

File: pkg/internal/controllers/test/selector.go
Comment (wallrj) [inline pkg/internal/controllers/test/selector.go:207]: Thanks for writing this new integration test. I observed the test fail

```console
$ git checkout origin/main -- pkg/internal/approver/manager/predicate/predicate.go

$ make test-unit
...
• [FAILED] [10.202 seconds]
Selector [It] it should select on CertificateRequests with issuerRef={name=my-issuer} where issuerRef={name=my-issuer kind=Issuer group=cert-manager.io}, RBAC bound, and plugin return Ready
/home/richard/projects/cert-manager/approver-policy/pkg/internal/controllers/test/selector.go:173

  Timeline >>
  STEP: Creating Policy Namespace:  @ 11/20/24 14:08:26.715
  STEP: Running Policy controller @ 11/20/24 14:08:26.716
  STEP: Waiting for Informers to Sync @ 11/20/24 14:08:26.716
  I1120 14:08:26.716954] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.CertificateRequest"
  I1120 14:08:26.716965] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1alpha1.CertificateRequestPolicy"
  I1120 14:08:26.716971] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.PartialObjectMetadata"
  I1120 14:08:26.716977] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.PartialObjectMetadata"
  I1120 14:08:26.716982] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.PartialObjectMetadata"
  I1120 14:08:26.716987] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.PartialObjectMetadata"
  I1120 14:08:26.716993] manager: Starting EventSource controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" source="kind source: *v1.PartialObjectMetadata"
  I1120 14:08:26.716997] manager: Starting Controller controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest"
  I1120 14:08:26.717022] manager: Starting EventSource controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" source="kind source: *v1alpha1.CertificateRequestPolicy"
  I1120 14:08:26.717035] manager: Starting EventSource controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" source="channel source: 0xc0002eb1f0"
  I1120 14:08:26.717044] manager: Starting Controller controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy"
  I1120 14:08:26.822172] manager: Starting workers controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" worker count=1
  I1120 14:08:26.822247] manager: Reconciling controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" CertificateRequestPolicy="selector-22blm" namespace="" name="selector-22blm" reconcileID="cc15212c-0062-401f-8d18-d28703ddb581"
  I1120 14:08:26.822274] controllers/certificaterequestpolicies: syncing name="selector-22blm"
  I1120 14:08:26.822315] controllers/certificaterequestpolicies: ready for approval evaluation name="selector-22blm" ready=true
  I1120 14:08:26.822506] manager: Starting workers controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" worker count=1
  I1120 14:08:26.822631] manager/events: CertificateRequestPolicy is ready for approval evaluation type="Normal" object={"kind":"CertificateRequestPolicy","name":"selector-22blm","uid":"f6eb38bc-7830-41dd-88df-04301c2881b3","apiVersion":"policy.cert-manager.io/v1alpha1","resourceVersion":"773"} reason="Ready"
  I1120 14:08:26.827795] manager: Reconcile successful controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" CertificateRequestPolicy="selector-22blm" namespace="" name="selector-22blm" reconcileID="cc15212c-0062-401f-8d18-d28703ddb581"
  I1120 14:08:26.828010] manager: Reconciling controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" CertificateRequestPolicy="selector-22blm" namespace="" name="selector-22blm" reconcileID="a886593d-87a6-4bf1-a015-f758be51306b"
  I1120 14:08:26.828041] controllers/certificaterequestpolicies: syncing name="selector-22blm"
  I1120 14:08:26.828061] controllers/certificaterequestpolicies: ready for approval evaluation name="selector-22blm" ready=true
  I1120 14:08:26.828256] manager/events: CertificateRequestPolicy is ready for approval evaluation type="Normal" object={"kind":"CertificateRequestPolicy","name":"selector-22blm","uid":"f6eb38bc-7830-41dd-88df-04301c2881b3","apiVersion":"policy.cert-manager.io/v1alpha1","resourceVersion":"775"} reason="Ready"
  I1120 14:08:26.831092] manager: Reconcile successful controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy" CertificateRequestPolicy="selector-22blm" namespace="" name="selector-22blm" reconcileID="a886593d-87a6-4bf1-a015-f758be51306b"
  I1120 14:08:26.907648] manager: Reconciling controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" CertificateRequest="test-policy-mwnbh/test-rlq9v" namespace="test-policy-mwnbh" name="test-rlq9v" reconcileID="45e453c3-925e-4250-9429-33856a777531"
  I1120 14:08:26.907662] controllers/certificaterequests: syncing certificaterequest namespace="test-policy-mwnbh" name="test-rlq9v"
  I1120 14:08:26.907684] controllers/certificaterequests: request was unprocessed namespace="test-policy-mwnbh" name="test-rlq9v"
  I1120 14:08:26.907709] manager: Reconcile successful controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest" CertificateRequest="test-policy-mwnbh/test-rlq9v" namespace="test-policy-mwnbh" name="test-rlq9v" reconcileID="45e453c3-925e-4250-9429-33856a777531"
  I1120 14:08:26.907725] manager/events: Request is not applicable for any policy so ignoring type="Normal" object={"kind":"CertificateRequest","namespace":"test-policy-mwnbh","name":"test-rlq9v","uid":"40d470bf-f46d-4830-a6e8-014cae11b8bb","apiVersion":"cert-manager.io/v1","resourceVersion":"781"} reason="Unprocessed"
  [FAILED] in [It] - /home/richard/projects/cert-manager/approver-policy/pkg/internal/controllers/test/util.go:57 @ 11/20/24 14:08:36.908
  I1120 14:08:36.908442] manager: Stopping and waiting for non leader election runnables
  I1120 14:08:36.908455] manager: Stopping and waiting for leader election runnables
  I1120 14:08:36.908475] manager: Shutdown signal received, waiting for all workers to finish controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest"
  I1120 14:08:36.908496] manager: Shutdown signal received, waiting for all workers to finish controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy"
  I1120 14:08:36.908503] manager: All workers finished controller="certificaterequestpolicy" controllerGroup="policy.cert-manager.io" controllerKind="CertificateRequestPolicy"
  I1120 14:08:36.908537] manager: All workers finished controller="certificaterequest" controllerGroup="cert-manager.io" controllerKind="CertificateRequest"
  I1120 14:08:36.908579] manager: Stopping and waiting for caches
  I1120 14:08:36.909086] manager: Stopping and waiting for webhooks
  I1120 14:08:36.909096] manager: Stopping and waiting for HTTP servers
  I1120 14:08:36.909102] manager: Wait completed, proceeding to shutdown the manager
  STEP: Waiting for all CertificateRequest resources to be deleted @ 11/20/24 14:08:36.909
  STEP: Deleting: test-policy-mwnbh/test-rlq9v @ 11/20/24 14:08:36.911
  STEP: Waiting for all CertificateRequestPolicy resources to be deleted @ 11/20/24 14:08:36.914
  << Timeline

  [FAILED] Timed out after 10.001s.
  expected approval
  Expected
      <bool>: false
  to be true
  In [It] at: /home/richard/projects/cert-manager/approver-policy/pkg/internal/controllers/test/util.go:57 @ 11/20/24 14:08:36.908
------------------------------
...
Summarizing 1 Failure:
  [FAIL] Selector [It] it should select on CertificateRequests with issuerRef={name=my-issuer} where issuerRef={name=my-issuer kind=Issuer group=cert-manager.io}, RBAC bound, and plugin return Ready
  /home/richard/projects/cert-manager/approver-policy/pkg/internal/controllers/test/util.go:57

Ran 50 of 50 Specs in 216.230 seconds
FAIL! -- 49 Passed | 1 Failed | 0 Pending | 0 Skipped
...
```

It's very difficult to understand why the test failed except by searching for line 57 of util.go and noticing that it is the `waitForApproval` function.

Comment (wallrj) [reply pkg/internal/controllers/test/selector.go:207]: Suggest removing the last few lines of the test, because it seems redundant. 
It seems to additionally test that the CRs referencing another cert-manager Issuer (not bound to a policy) do not get approved. But why? 

```go
		crName = createCertificateRequest(ctx, env.UserClient, namespace.Name, gen.SetCSRDNSNames(),
			gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{Name: "my-issuer-2"}),
		)
		waitForNoApproveOrDeny(ctx, env.AdminClient, namespace.Name, crName)
```

I suggest removing it or adding comments or `By(...)` lines explaining the purpose of this part of the test.

**Review by wallrj — APPROVED** [review]
Thanks @erikgb 

I left a few comments and suggestions below.

Address or answer those and unhold and merge at will.

/approve
/lgtm
/hold in case you need to make changes.

File: pkg/internal/controllers/test/selector.go
Comment (erikgb) [reply pkg/internal/controllers/test/selector.go:203]: I have now removed the lines. It seems like you have hit a flake with the failing test. I haven't seen it failing while working on this. But I have seen flakes in CI in this project previously. I might take a look to see if I can find the culprit, but that's anyway not related to this change. PTAL!

**Comment by erikgb** [comment]
Thanks for reviewing, @wallrj!

/unhold

**Pull request merged**
